infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+
+  private static final Pattern REFERENCE_PATTERN = Pattern.compile("\\$\\(\\w+\\)");
+
+  private final Map<String, SortedSet<String>> referenceMap;
+  private final Map<String, SortedSet<String>> recursiveReferenceMap;
+
+  private EnvironmentVariablesSorter(Map<String, EnvVar> envVars) {
+    this.referenceMap = new HashMap<>(envVars.size());
+    for (Map.Entry<String, EnvVar> e : envVars.entrySet()) {
+      referenceMap.put(e.getKey(), findReferences(e.getValue()));
+    }
+
+    recursiveReferenceMap = new HashMap<>(envVars.size());
+  }
+
+  /**
+   * Sorts the provided env variables so that a variable depending on some other variable is always
+   * later in the returned list than the env var it depends on.
+   *
+   * @param envVars the collection of environment variables to sort
+   * @return a sorted list of the environment variables
+   * @throws InfrastructureException if there is a cycle in the value dependencies and thus the
+   *     expansion is impossible
+   */
+  static List<EnvVar> sort(Map<String, EnvVar> envVars) throws InfrastructureException {
+    try {
+      EnvironmentVariablesSorter sorter = new EnvironmentVariablesSorter(envVars);
+      List<EnvVar> ret = new ArrayList<>(envVars.values());
+      ret.sort(sorter);
+      return ret;
+    } catch (CycleException e) {
+      throw new InfrastructureException(e.getMessage());
+    }
+  }
+
+  @Override
+  public int compare(EnvVar a, EnvVar b) throws CycleException {
+    if (a == b) {
+      return 0;
+    }
+
+    SortedSet<String> aRefs = referenceMap.get(a.getName());
+    SortedSet<String> bRefs = referenceMap.get(b.getName());
+
+    // quickly deal with the common case of no references in the value
+    if (aRefs.isEmpty()) {
+      // we need to have a total ordering, so if there is no difference in references, we need
+      // to pick a different sorting criteria
+      return bRefs.isEmpty() ? a.getName().compareTo(b.getName()) : -1;
+    } else if (bRefs.isEmpty()) {
+      // b has no references, a has, therefore a > b.
+      return 1;
+    }
+
+    // ok, we have variable references in both a and b
+    // we need to establish the transitive closure over the references to figure out whether there
+    // is a cycle in the dependencies, or not
+    SortedSet<String> aTransRefs = transitiveClosure(a.getName());
+    SortedSet<String> bTransRefs = transitiveClosure(b.getName());
+
+    // we don't need to check for cycle here, because that has been detected when establishing
+    // the closure.
+    if (aTransRefs.contains(b.getName())) {
+      // a depends on b, hence a > b.
+      return 1;
+    } else if (bTransRefs.contains(a.getName())) {
+      // b depends on a -> a < b.
+      return -1;
+    }
+
+    // ok, the two env vars don't depend on each other, so we actually don't care too much about
+    // the order. We need to be consistent and total, though...
+
+    int sizeDiff = aTransRefs.size() - bTransRefs.size();
```

Comment:
You're right as long as we keep the rest of the code as is (-ish) and don't make this class reusable. If it could be used outside of the `sort` method, I would try to keep the implementation conform to the general contract of comparator - it needs to be consistent with equals and be total, which we'd be breaking with your suggestion.

Reply:
I didn't mean to make it reusable, see my comments above. And sorry for the confusion. From what I saw it is unusual to sort variables in an alphabetical order

Reply:
I'm still a little bit nervous about breaking the contract of comparable here. I didn't try it yet, but if we just used the dependency as the sorting criteria, the following is easily `true`:

```
a == b && b == c && a > c
```

(where `==` and `<` are defined in terms of our comparator)

E.g.
```
A="$(C)"
B="blah"
C="bleh"
```

I am no expert on the inner workings of the "stable, adaptive iterative mergesort" used by `List.sort` but I can easily imagine the above conditions producing different (and sometimes incorrect) order under different initial conditions (meaning the original order of the elements).

The comparator should by contract be "total" which amongst other things also means "transitive" as far as I looked. The above example breaks the transitivity assumption.

Reply:
I'm not a sorting expert either, but the current approach seems overcomplicated. But the violation of sorting prerequisites can break the algorithm indeed, so should be present.